src/csharp/Grpc.Core.Tests/Interceptors/ClientInterceptorTest.cs

```diff
@@ -0,0 +1,76 @@
+#region Copyright notice and license
+
+// Copyright 2017 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core;
+using Grpc.Core.Interceptors;
+using Grpc.Core.Internal;
+using Grpc.Core.Utils;
+using Grpc.Core.Tests;
+using NUnit.Framework;
+
+namespace Grpc.Core.Interceptors.Tests
+{
+    public class ClientInterceptorTest
+    {
+        private class AddHeaderClientInterceptor : ClientInterceptor
+        {
+            readonly Metadata.Entry header;
+            public AddHeaderClientInterceptor(string key, string value)
+            {
+                this.header = new Metadata.Entry(key, value);
+            }
+            public override TResponse BlockingUnaryCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request, Func<Method<TRequest, TResponse>, string, CallOptions, TRequest, TResponse> next)
+            {
+                options.Headers.Add(this.header);
+                return next(method, host, options, request);
+            }
+
+            public Metadata.Entry Header
+            {
+                get
+                {
+                    return this.header;
+                }
+            }
+        }
+
+        const string Host = "127.0.0.1";
+
+        [Test]
+        public void AddRequestHeaderInClientInterceptor()
+        {
+            var helper = new MockServiceHelper(Host);
+            var interceptor = new AddHeaderClientInterceptor("x-client-interceptor", "hello world");
+            helper.UnaryHandler = new UnaryServerMethod<string, string>((request, context) =>
+            {
+                var interceptorHeader = context.RequestHeaders.Last(m => (m.Key == interceptor.Header.Key)).Value;
+                Assert.AreEqual(interceptorHeader, interceptor.Header.Value);
+                return Task.FromResult("PASS");
+            });
+            var server = helper.GetServer();
+            server.Start();
+            var callInvoker = helper.GetChannel().Intercept(interceptor);
```

Comment:
Yes, that is how it is going to work.

I am considering adding another overload that takes `params Interceptor[]` and adds the list in-order (not the opposite order implied by decorator style).

Reply:
I'd definitely add this order of operations to be clear. I pref FIFO, but I would say you want both functions to behave the same versus supporting one the opposite way.

Reply:
I think FIFO does not make sense when chaining calls like `.Intercept(a).Intercept(b)`, because you can imagine a scenario in which you are given a `CallInvoker` by another part of your program that is created with `channel.Intercept` and you then decide to intercept it again and see the whole thing as a black box.

However, I do agree that an overload that takes multiple interceptors in a single call: `.Intercept(a, b, c)` should pass control to `a`, `b`, and `c` in order.

Reply:
I think the Intercept(a, b, c) overload would be useful.

Reply:
I think it'd be nice to also have an overload like:
```
.Intercept(p => {
        p.AddLogging();
        p.AddAuth();
    });
```
where p is a pipeline object of some sort.

This would make it very easy to develop packagable pipeline plugins by just creating an extension method for the pipeline object. Additionally in cases where you don't about the type of TRequest / TResponse (eg. logging) you could create a generic extension method so that your plugin will work for all types of requests / responses.

Reply:
@plaisted Can you clarify what is the advantage of that over

```csharp
.Intercept(new Logger(), new Authenticator());
```

?

Reply:
@mehrdada 

Functionally they are the same but using the extension method allows you to nicely encapsulate what "API" you want the user of the package to use vs. what the actual constructor that the class needs.

As a contrived example the `Authenticator` interceptor may require a `TokenValidator` service so it's constructor is actually `Authenticator(ITokenValidator validator)`. Adding an extensions method in your package of `AddJwtAuth()` would simply construct the `Authenticator` interceptor with a `JwtTokenValidator` that implements `ITokenValidator`.

Now the extension method would generally do a lot more than that, maybe even add multiple interceptors which is trivial with this sort of setup:
```csharp
//extension method
public static void AddJwtAuth(this IInterceptorPipeline pipeline, Action<JwtOptions> optionsBuilder) {
    var options = optionsBuilder(new JwtOptions());
    pipeline.Add(new Authenticator(new JwtTokenValidator(options)));
    pipeline.AddInterceptor2(); // add a dependent interceptor
    pipeline.AddInterceptor3(); // add another
}

//using it
.Intercept(p => {
        p.AddJwtAuth(o=>{ o.ValidationCertificate = cert; });
    });
```

I've written a few addons (aspcore mainly) that use this sort of approach and it works very nicely as you don't have to expose any implementation details to the users of the package.  Admittedly all of this other than adding multiple interceptors could all be implemented relatively easily using a static "factory" method so a lot of this is preference and wanting to stay similar to how I see this done across different frameworks in c#.